### ISSUE TITLE
DC-ASGD(Delay Compensated Asynchronous Stochastic Gradient Descent)?

### ISSUE THREAD

**Issue by liufei1656** [tensorflow/tensorflow#8744]
DC-ASGD is Microsoft's very useful algorithm for distributed asynchronous training. Compared with the ordinary ASGD algorithm, DC-ASGD has no significant loss in speed, but can get almost the same effect as Sequential SGD. As far as I know, other mainstream deep learning open source tools have implemented this algorithm, such as: CNTK, Mxnet, Paddle and so on. But in Tensorflow I have not found similar modules.

### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
[https://github.com/Microsoft/CNTK/issues/1295](https://github.com/Microsoft/CNTK/issues/1295)
[https://github.com/PaddlePaddle/Paddle/issues/185](https://github.com/PaddlePaddle/Paddle/issues/185)
[https://github.com/dmlc/mxnet/pull/3614](https://github.com/dmlc/mxnet/pull/3614)

### What other attempted solutions have you tried?
I tried to implement this algorithm in Tensorflow by myself.  I do not have enough ability to do this now.

### The link address of the paper
[Asynchronous Stochastic Gradient Descent with Delay Compensation for Distributed Deep Learning](https://arxiv.org/abs/1609.08326)

**Comment by renyi533** [comment]
I am also very interested with this. Any update?

**Comment by alisidd** [comment]
I can do this

**Comment by alisidd** [comment]
@aselle Tensorflow's documentation on distributed computing leads me to believe that the end user is responsible for assigning workers and parameter servers to different devices. Should I then only implement Gradient Descent with Delay Compensation, which the user will have to implement asynchronously?

**Comment by alisidd** [comment]
@aselle Tensorflow's documentation on distributed computing leads me to believe that the end user is responsible for assigning workers and parameter servers to different devices. Should I then only implement Gradient Descent with Delay Compensation, which the user will have to implement asynchronously?

**Comment by aselle** [comment]
@mrry, might be able to give you more direction in this vein (also @tfboyd )

**Comment by mrry** [comment]
I haven't read the paper in depth, but from a quick skim of the paper you could implement the update rule as a `tf.train.Optimizer` subclass, and the asynchronous execution would be provided by whatever training loop the user used. (It would presumably "work" in a synchronous case as well, because *w<sub>t</sub>* &ndash; *w<sub>bak</sub>* would be zero, and hence it would devolve into classic SGD.)

**Comment by YirenBj** [comment]
I already implemented this. Currently, I am verifying its impact on metrics, such as AUC, in our ads data. I will send the pull request if this is really useful.

**Comment by MorganGellert** [comment]
Is support for the adaptive variance parameter version of DC-ASGD planned on being added? The paper found that it performed better in all cases than the constant version which is currently implemented.

**Comment by dynamicwebpaige** [comment]
Closing this issue, as the PR for DC-ASGD was merged ([#9551](https://github.com/tensorflow/tensorflow/pull/9551)). Thank you! 👍
